src/query/subquery_cache.c

```diff
@@ -0,0 +1,749 @@
+ *   xasl(in): The XASL node for which all cache entries will be cleared.
+ *
+ * This function clears all cache entries associated with a given XASL node. It does so by iterating over the hash table and
+ * freeing all keys and values. This function is typically called when resetting or destroying the cache for a XASL node.
+ */
+void
+sq_cache_drop_all (xasl_node * xasl)
+{
+  if (xasl->sq_cache_flag & SQ_CACHE_INITIALIZED_FLAG)
+    {
+      if (xasl->sq_cache_ht != NULL)
+	{
+	  mht_clear (xasl->sq_cache_ht, sq_rem_func, NULL);
+	}
+    }
+}
+
+/*
+ * sq_cache_destroy () - Destroys the cache for a given XASL node.
+ *   xasl(in): The XASL node for which the cache is being destroyed.
+ *
+ * This function destroys the cache associated with a given XASL node. It clears all cache entries and then destroys the hash
+ * table itself. It also resets cache-related flags and counters for the XASL node. This function is called when a XASL node is
+ * no longer needed or before it is deallocated.
+ */
+void
+sq_cache_destroy (xasl_node * xasl)
+{
+  if (xasl->sq_cache_flag & SQ_CACHE_INITIALIZED_FLAG)
+    {
+      er_log_debug (ARG_FILE_LINE,
+		    "destroy sq_cache at xasl %p\ncache info : \n\thit : %10lu\n\tmiss: %10lu\n\tsize: %10lu Bytes\n",
+		    xasl, xasl->sq_cache_hit, xasl->sq_cache_miss, xasl->sq_cache_size);
+      sq_cache_drop_all (xasl);
+      mht_destroy (xasl->sq_cache_ht);
+    }
+  xasl->sq_cache_flag = 0;
+  xasl->sq_cache_ht = NULL;
+}
+
+/*
+ * execute_regu_variable_xasl_with_sq_cache () - Executes a regu variable XASL with support for scalar subquery result caching.
+ *   return: False if execution should proceed after caching logic, True otherwise.
+ *   thread_p(in): Thread context.
+ *   regu_var(in): The regu variable to be executed.
+ *   vd(in): Value descriptor for parameter values.
+ *
+ * This function attempts to execute a regu variable's XASL with caching logic for scalar subquery results. It checks if the
+ * XASL node associated with the regu variable is eligible for caching and either retrieves the cached result or executes the
+ * XASL and caches the result if appropriate. This function aims to improve performance by avoiding redundant executions of
+ * scalar subqueries.
+ */
+int
+execute_regu_variable_xasl_with_sq_cache (THREAD_ENTRY * thread_p, REGU_VARIABLE * regu_var, void *vd)
+{
+  if (regu_var->xasl && !(regu_var->xasl->sq_cache_flag & SQ_CACHE_NOT_CACHING_FLAG)
+      && (regu_var->xasl->status == XASL_CLEARED || regu_var->xasl->status == XASL_INITIALIZED))
+    {
+
+      if (regu_var->xasl->sq_cache_flag & SQ_CACHE_ENABLED_FLAG)
+	{
+	  if (sq_get (regu_var->xasl, regu_var))
+	    {
+	      regu_var->xasl->status = XASL_SUCCESS;
+	      return true;
+	    }
+	  EXECUTE_REGU_VARIABLE_XASL (thread_p, regu_var, (val_descr *) vd);
+	  if (CHECK_REGU_VARIABLE_XASL_STATUS (regu_var) != XASL_SUCCESS)
+	    {
+	      return false;
+	    }
+	  sq_put (regu_var->xasl, regu_var);
```

Comment:
xasl was executed because there were no cached values.
But should we check once more whether the same key exists by calling `mht_put_if_not_exists` on `sq_put`?